src/bot/exts/dragonfly/dragonfly.py

```diff
@@ -261,6 +262,199 @@ async def report(self: Self, interaction: discord.Interaction, button: discord.u
             await interaction.edit_original_response(view=self)
 
 
+class NoteModal(discord.ui.Modal, title="Add a note"):
+    """A modal that allows users to add a note to a package."""
+
+    _interaction: discord.Interaction | None = None
+    note_content = discord.ui.TextInput(
+        label="Content",
+        placeholder="Enter the note content here",
+        min_length=1,
+        max_length=1000,  # Don't want to overfill the embed
+    )
+
+    def __init__(self, embed: discord.Embed, view: discord.ui.View) -> None:
+        super().__init__()
+
+        self.embed = embed
+        self.view = view
+
+    async def on_submit(self, interaction: discord.Interaction) -> None:
+        """Modal submit callback."""
+        if not interaction.response.is_done():
+            await interaction.response.defer()
+        self._interaction = interaction
+
+        content = f"{self.note_content.value} • {interaction.user.mention}"
+
+        # We need to check what fields the embed has to determine where to add the note
+        # If the embed has no fields, we add the note and return
+        # Otherwise, we need to make sure the note is added after the event log
+        # This involves clearing the fields and re-adding them in the correct order
+        # Which is why we save the event log in a variable
+
+        match len(self.embed.fields):
+            case 0:  # Package is awaiting triage, no notes or event log
+                notes = [content]
+                event_log = None
+            case 1:  # Package either has notes or event log
+                if self.embed.fields[0].name == "Notes":
+                    notes = [self.embed.fields[0].value, content]
+                else:
+                    event_log = self.embed.fields[0].value
+                    notes = [content]
+                self.embed.clear_fields()
+            case 2:  # Package has both notes and event log
+                if self.embed.fields[0].name == "Notes":
+                    notes = [self.embed.fields[0].value, content]
+                    event_log = self.embed.fields[1].value
+                else:
+                    notes = [self.embed.fields[1].value, content]
+                    event_log = self.embed.fields[0].value
+                self.embed.clear_fields()
```

Comment:
This part feels kind of messy, can we abstract it away into another function?

Reply:
I couldn't think of any other logic to abstract this, neither did I want to move into it's own function because its quite intertwined with the variables in the local scope and is only called once. Would it be possible to leave it be?

Reply:
yeah, should be fine, it's functional